### PR TITLE
Track nullability in script code

### PR DESCRIPTION
**Pull request by filipw**
At the moment script code is explicitly excluded from nullability analysis.

So the following script doesn't produce any warnings:

```
#nullable enable
string foo = null
```

This PR includes script code into the nullability analysis.
Additionally, since scripting doesn't expose an API to set Language Version, I bumped it to C# 8 (since it should always be the newest). Once C# 8 ships, we could move it back to `Latest`.

### CODE REVIEW

**Comment by filipw** [comment]
rebased this on top of master

File: src/Scripting/CSharp/CSharpScriptCompiler.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ internal sealed class CSharpScriptCompiler : ScriptCompiler
     {
         public static readonly ScriptCompiler Instance = new CSharpScriptCompiler();
 
-        private static readonly CSharpParseOptions s_defaultOptions = new CSharpParseOptions(kind: SourceCodeKind.Script, languageVersion: LanguageVersion.Latest);
+        private static readonly CSharpParseOptions s_defaultOptions = new CSharpParseOptions(kind: SourceCodeKind.Script, languageVersion: LanguageVersion.CSharp8);
```
Comment (cston) [inline src/Scripting/CSharp/CSharpScriptCompiler.cs]: >CSharp8 [](start = 155, length = 7)

Adding @jaredpar, @tmat to review version number change.

Comment (jaredpar) [reply src/Scripting/CSharp/CSharpScriptCompiler.cs]: Can't take the version change here. The script and C# compiler are in sync with respect to versions. The script compiler will move to eight at the same time Latestn moves to 8.

Comment (tmat) [reply src/Scripting/CSharp/CSharpScriptCompiler.cs]: Yes, I think it makes sense to allow specify language version in ScriptOptions. An interesting question is whether different submissions in the chain can use different C# language versions. I think it should be ok. Might be actually useful, if users can do something like `#langversion 8.0` in Interactive Window (as a host command) to switch to a different version to try how some snippet works in that version.

Also the VS C# Interactive Window should switch to max version like we did for EEs.

---
In reply to: [254033944](https://github.com/dotnet/roslyn/pull/33096#discussion_r254033944) [](ancestors = 254033944)

**Comment by filipw** [comment]
@tmat I have updated the PR so that scripting uses `LanguageVersion.Latest` again, like @jaredpar requested.
I also came up with a solution of setting `LanguageVersion` via `ScriptOptions` which I think makes sense and - as we already discussed here - could be quite useful. It also enables nullability analysis for scripting.
If you have some time, let me know if we could go forward with such approach - thanks!

File: src/Scripting/CSharp/ScriptOptionsExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.CodeAnalysis.CSharp.Scripting
+{
+    public static class ScriptOptionsExtensions
+    {
+        public static ScriptOptions WithLanguageVersion(this ScriptOptions options, LanguageVersion languageVersion)
+        {
+            if (options.ParseOptions is CSharpParseOptions csharpParseOptions && csharpParseOptions.LanguageVersion == languageVersion)
+                return options;
```
Comment (tmat) [inline src/Scripting/CSharp/ScriptOptionsExtensions.cs]: I think it would be better to throw here if `options.ParseOptions` is non-null and not C# parse options. 
Consider we more language specific options. Then setting one of the options of one language would drop all of the options set by the other language.
Something like:

```C#
var parseOptions = 
  (options.ParseOptions is null) ? CSharpScriptCompiler.DefaultParseOptions :
  (options.ParseOptions is CSharpParseOptions existing) ? existing : 
  throw new InvalidOperationException(...));

return options.WithParseOptions(parseOptions.WithLanguageVersion(languageVersion)
```

File: src/Scripting/Core/ScriptCompiler.cs [hunk 3]
```diff
@@ -13,7 +13,7 @@ internal abstract class ScriptCompiler
         public abstract DiagnosticFormatter DiagnosticFormatter { get; }
         public abstract StringComparer IdentifierComparer { get; }
 
-        public abstract SyntaxTree ParseSubmission(SourceText text, CancellationToken cancellationToken);
+        public abstract SyntaxTree ParseSubmission(SourceText text, ScriptOptions scriptOptions, CancellationToken cancellationToken);
```
Comment (tmat) [inline src/Scripting/Core/ScriptCompiler.cs]: >ScriptOptions [](start = 68, length = 13)

I'd type this to `ParseOptions` since the parsing should not depend on other options.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.cs [hunk 4]
```diff
@@ -105,7 +105,7 @@ internal Binder GetBinder(SyntaxNode node, CSharpSyntaxNode memberDeclarationOpt
             // Unless this is interactive retrieving a binder for global statements
             // at the very top-level (i.e. in a completely empty file) use
             // node.Parent to maintain existing behavior.
-            if (!InScript || node.Kind() != SyntaxKind.CompilationUnit)
+            if ((!InScript || node.Kind() != SyntaxKind.CompilationUnit) && node.Parent != null)
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: >node.Parent != null [](start = 76, length = 19)

When is this additional check needed?

Comment (filipw) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: There are over a 100 tests that fail inside `NullableWalker` without that extra check. You can see them here on one of the earlier CI runs for this PR https://dev.azure.com/dnceng/public/_build/results?buildId=87553&view=ms.vss-test-web.build-test-results-tab

They were mentioned here too https://github.com/dotnet/roslyn/pull/33093#discussion_r253562102

File: src/Scripting/CSharp/ScriptOptionsExtensions.cs (hunk 2 above)
Comment (filipw) [reply src/Scripting/CSharp/ScriptOptionsExtensions.cs]: good idea, I added this. not sure what the localization process for errors is though, because the error message is now in the resources file and exists only in English

File: src/Scripting/CSharp/CSharpScriptingResources.resx [hunk 5]
```diff
@@ -124,21 +124,24 @@
     <value>Copyright (C) Microsoft Corporation. All rights reserved.</value>
   </data>
   <data name="InteractiveHelp" xml:space="preserve">
-    <value><![CDATA[Usage: csi [option] ... [script-file.csx] [script-argument] ...
```
Comment (cston) [inline src/Scripting/CSharp/CSharpScriptingResources.resx]: ><CDATA[ [

Please consider reverting the `<![CDATA[` change if it's not needed.

Comment (filipw) [reply src/Scripting/CSharp/CSharpScriptingResources.resx]: sorry, my bad - it happened automatically in VS - will revert 👍

**Comment by filipw** [comment]
I rebased this on top of master and resolved conflicts that appearee due one of the recent merges.

File: src/Scripting/CSharp/CSharpScriptingResources.resx (hunk 5 above)
Comment (cston) [inline src/Scripting/CSharp/CSharpScriptingResources.resx]: ><CDATA[ [

Please consider reverting this change.

File: src/Scripting/CSharpTest/ScriptOptionsTests.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        [Fact]
+        public void WithLanguageVersion()
+        {
+            var options = ScriptOptions.Default.WithLanguageVersion(LanguageVersion.CSharp8);
+            Assert.Equal(LanguageVersion.CSharp8, ((CSharpParseOptions)options.ParseOptions).LanguageVersion);
+        }
+
+        [Fact]
+        public void WithWithLanguageVersion_SameValueTwice_DoesNotCreateNewInstance()
```
Comment (cston) [inline src/Scripting/CSharpTest/ScriptOptionsTests.cs]: >WithWith [](start = 20, length = 8)

Typo, here and following test?

File: src/Scripting/CSharp/CSharpScriptingResources.resx (hunk 5 above)
Comment (filipw) [reply src/Scripting/CSharp/CSharpScriptingResources.resx]: fixed now 🤦‍♂️

File: src/Scripting/CSharp/CSharpScriptingResources.resx [hunk 7]
```diff
@@ -125,9 +125,7 @@
   </data>
   <data name="InteractiveHelp" xml:space="preserve">
     <value><![CDATA[Usage: csi [option] ... [script-file.csx] [script-argument] ...
-
 Executes script-file.csx if specified, otherwise launches an interactive REPL (Read Eval Print Loop).
```
Comment (cston) [inline src/Scripting/CSharp/CSharpScriptingResources.resx]: Blank lines were removed.

Comment (filipw) [reply src/Scripting/CSharp/CSharpScriptingResources.resx]: force pushed again - sorry for the trouble

File: src/Scripting/CSharp/CSharpScriptingResources.resx [hunk 8]
```diff
@@ -141,4 +141,7 @@ Options:
   --                             Indicates that the remaining arguments should not be treated as options.
 ]]></value>
   </data>
+  <data name="CannotSetCSharpLanguageVersion" xml:space="preserve">
+    <value>Cannot set C# LanguageVersion because ParseOptions were already configured for a different language.</value>
```
Comment (tmat) [inline src/Scripting/CSharp/CSharpScriptingResources.resx]: >ParseOptions [](start = 49, length = 12)

The fact that we use ParseOptions is an implementation detail, so the user wouldn't know from the message what's going on here. 
I would reword this to 

```
Cannot set {0} specific option {1} because the script was already configured for a different language.
```

where {0} would be `C#` and {1} would be `nameof(LanguageVersion)`. Parameterizing the message makes it easier to translate (non-translated words are not part of the resource string) and also allows us to use the message for other language specific options if needed.

Comment (filipw) [reply src/Scripting/CSharp/CSharpScriptingResources.resx]: done. given this approach, I also moved the message from `CSharpScriptingResources.resx` to `ScriptingResources.resx`

**Comment by filipw** [comment]
I also noticed - unrelated to this PR - a formatting warning IDE0055 [on this line](https://github.com/dotnet/roslyn/blob/master/src/Scripting/CoreTest/ScriptOptionsTests.cs#L297), which was introduced by https://github.com/dotnet/roslyn/pull/32668, so I fixed that too.

**Comment by filipw** [comment]
@jaredpar is this OK for you?

File: src/Scripting/CSharp/CSharpScriptingResources.Designer.cs [hunk 9]
```diff
@@ -20,7 +19,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Scripting {
     // class via a tool like ResGen or Visual Studio.
     // To add or remove a member, edit your .ResX file then rerun ResGen
     // with the /str option, or rebuild your VS project.
-    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "4.0.0.0")]
+    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "16.0.0.0")]
```
Comment (jaredpar) [inline src/Scripting/CSharp/CSharpScriptingResources.Designer.cs]: This change looks suspicious but I also don't think it's wrong.

**Comment by filipw** [comment]
I resolved the conflicts after #29069 was merged. The integration test seems flaky (failure unrelated to the PR).
Given that everyone signed off, is it OK to get this merged?

**Comment by cston** [comment]
Thanks @filipw. Sorry for the delay.

Would you mind if I squash commits when merging?

**Comment by filipw** [comment]
> Would you mind if I squash commits when merging?

sure no problem, it's gotten a bit messy indeed 😅

**Comment by cston** [comment]
Thanks @filipw for the PR!
